Generalize evolution actions to arbitrary temporal_ids

nilsvu opened the pull request:
## Proposed changes

This PR adds a `step_prefix` template alias to the tags that may be used as a `temporal_id`:

- `Tags::TimeId` uses `Tags::dt`
- `LinearSolver::Tags::IterationId` uses `LinearSolver::Tags::OperatorAppliedTo`

This `step_prefix` is used by `ComputeTimeDerivative` (for evolution systems, renamed from `ComputeVolumeDuDt`), `ComputeOperatorAction` (for elliptic systems) and `ApplyFluxes` (for both, renamed from `ApplyBoundaryFluxesGlobalTimestepping`) to prefix the `variables_tag` for storing the results of their computations.

I also started to add documentation to actions on what types they require of the `Metavariables` and `system`. The types are explained in more detail in the `Evolution/Elliptic systems` Doxygen groups. I suggest we keep improving on this documentation in the future.

**What breaks**:
- `Actions::ComputeVolumeDuDt ` must be renamed to `Actions::ComputeTimeDerivative`
- `dg::Actions::ApplyBoundaryFluxesGlobalTimestepping ` must be renamed to `dg::Actions::ApplyFluxes`
- `system::du_dt` must be renamed to `system::compute_time_derivative`.

### Types of changes:

- [x] New feature

### Component:

- [x] Code
- [x] Documentation

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add get_variables_tags_list helper function

nilsvu commented:
It's probably best to look at the commits one-by-one. The last commit touches a number of files, but it only propagates the changes from the first commits to the codebase.

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp:
```diff
@@ -31,57 +31,69 @@ struct Mesh;
-
     using system = typename Metavariables::system;
-    constexpr size_t volume_dim = system::volume_dim;
+    constexpr size_t Dim = system::volume_dim;
```
we use `snake_case` for variable names, `dim` would be preferred :)

wthrowe replied:
I don't see a point in changing it from the original `volume_dim`.

nilsvu replied:
agreed

nilsdeppe commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -9,7 +9,6 @@
 #include "Domain/DomainCreators/RegisterDerivedWithCharm.cpp"
```
I think this commit needs to be squashed with the second once review is done so each individual commit compiles, right?

nilsdeppe reviewed (commented):
Just one minor stylistic comment that I'm happy for you to squash.

@wthrowe @kidder could you please look at this when you get a chance? Thanks!

wthrowe commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -598,6 +598,30 @@ struct variables_tag_with_tags_list_impl<DispatchTagType::Prefix> {
   using f = typename helper<Tag, NewTagsList>::type;
 };
+
+// Implementation of get_variables_tags_list_impl
```
Remove "_impl".

wthrowe commented on docs/GroupDefs.hpp:
```diff
@@ -397,8 +397,27 @@
+ * `db::add_tag_prefix<Metavariables::temporal_id::step_prefix, variables_tag>`
+ * (which represent the linear operator applied to the variables) as not-null
+ * pointers.
+ *   - The types of the `argument_tags` as constant references.
```
How are these last two supposed to be exposed by the system?  And what are `argument_tags` the arguments to?

nilsvu replied:
This refers to the volume computer struct, the list is indented

wthrowe replied:
Oops, I missed that.

wthrowe commented on docs/GroupDefs.hpp:
```diff
@@ -418,7 +437,30 @@
+ *   - The types of the tensors in
+ * `db::add_tag_prefix<Metavariables::temporal_id::step_prefix, variables_tag>`
+ * (which represent the time derivatives of the variables) as not-null pointers.
+ *   - The types of the `argument_tags` as constant references.
```
Same comments as above.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp:
```diff
@@ -0,0 +1,81 @@
+ *   - `temporal_id::step_prefix`
+ * - System:
+ *   - `variables_tag`
+ *   - `volume`
```
I don't like the name "volume" here.  There are a lot of things associated with the volume, this should indicate which one it is.  (In fact, this action doesn't know anything about DG, so the "volume" association isn't really right anyway, but SpECTRE is sufficiently DG-focused that I think that's OK.)

nilsvu replied:
Agreed, but I couldn’t think of a better name. Do you have a suggestion?

nilsvu replied:
Actually, everything that I can think of seems redundant (volume terms/contribution/operator/...). If what we mean with „system“ are the PDEs in some discretization, then this operator does indeed describe the volume part of it, so `system::volume` seems appropriate. Thoughts?

nilsleiffischer replied:
Actually, everything that I can think of seems redundant (volume terms/contribution/operator/...). If what we mean with „system“ are the PDEs in some discretization, then this operator does indeed describe the volume part of it, so `system::volume` seems appropriate. Thoughts?

wthrowe replied:
I agree that the "system" is the (discretized) PDEs, but PDEs don't have a volume or a volume part.  And actually I've reconsidered what I said above due to the discussion on moving this class and I don't want "volume" in this name at all.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp:
```diff
@@ -0,0 +1,81 @@
+ *   - `db::add_tag_prefix<step_prefix, variables_tag>`
+ */
+template <typename Metavariables>
+struct ComputeVolume {
```
Same comment as for the system tag.  We're not computing the volume itself, but the volume terms or similar.

nilsvu replied:
If we keep this in the `dg` namespace then this means the volume contribution to the dg discretisation. Only if we try to be too general with this does it become ambiguous I think. But if we really want to generalize we may as well go all the way and have an `Action::MutateApply` that works with any struct that has an appropriate interface for `db::mutate_apply`. So my proposal with this PR is to have this DG-specialized `dg::Actions::ComputeVolume` that retrieves what it needs from Metavariables/System. For more general applications we should add `Action:MutateApply`.

wthrowe replied:
A general action might be the right way to go here, yeah.

kidder replied:
Yes, @nilsdeppe and I have also discussed introducing an `Action::MutateApply` as this pattern occurs frequently.  If it allows us to avoid an ambiguous name in this case, maybe it is worth introducing now.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp:
```diff
@@ -31,57 +31,69 @@ struct Mesh;
+struct ApplyFluxes {
+  template <typename Tag>
+  using step_prefix =
+      typename Metavariables::temporal_id::template step_prefix<Tag>;
```
This is only used once.  If you inline it you can avoid templating the action.

nilsvu replied:
The compiler complained when I tried that so I had to go with the type alias, do you know how I can make it work inline?

wthrowe replied:
`using dt_variables_tag = db::add_tag_prefix<Metavariables::temporal_id::template step_prefix, variables_tag>`

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ComputeVolume.hpp:
```diff
@@ -0,0 +1,81 @@
+struct ComputeVolume {
+  template <typename Tag>
+  using step_prefix =
+      typename Metavariables::temporal_id::template step_prefix<Tag>;
```
This is only used once.  It you inline it you can avoid templating the action.

wthrowe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CMakeLists.txt:
```diff
@@ -3,9 +3,10 @@
   Actions/Test_ApplyBoundaryFluxesLocalTimeStepping.cpp
+  Actions/Test_ApplyFluxes.cpp
   Actions/Test_ComputeNonconservativeBoundaryFluxes.cpp
+  Actions/Test_ComputeVolume.cpp
```
This isn't DG-specific.  Don't move it into DG.

nilsvu replied:
Not sure what the distinction is between dg and non-dg. If this is a volume contribution to the dg discretization of the System it may as well be in dg. Where else should I move it (must be outside of evolution)?

wthrowe replied:
This action was designed to be usable for any ODE system, not just those derived from DG or even from PDEs.  We have intended non-PDE use cases for this action.  (Various ray-tracing problems, mostly.)

This is actually already done in the tests.  In moving this you made the "Time" tests depend on the `dg` namespace, which they should not need.

The "Volume" in the name was kind of inappropriate, and I'd actually intended to drop it whenever I (or someone else) added the first non-DG system.

wthrowe commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -598,6 +598,30 @@ struct variables_tag_with_tags_list_impl<DispatchTagType::Prefix> {
+}  // namespace DataBox_detail
+
+template <typename Tag>
+using get_variables_tags_list =
```
Needs a test.

nilsvu committed:
Generalize evolution actions to arbitrary temporal_ids

A temporal_id must now expose a step_prefix, this is just Tags::dt
for time-stepping in evolution code. The DG operators use it to
prefix the variables_tag for storing the results of their computations.
The generalization allows the elliptic code to use the same actions,
since time-stepping is replacecd by linear solver iterations.
Also note that system::du_dt is renamed to compute_time_derivative.

nilsvu commented:
Refactored in response to our discussion. We now have `ComputeTimeDerivative` for evolution systems, `ComputeOperatorAction` for elliptic systems and `ApplyFluxes` for both.

nilsvu commented:
@wthrowe @nilsdeppe @kidder this should be ready

nilsdeppe commented on tests/Unit/DataStructures/DataBox/Test_DataBoxTag.cpp:
```diff
@@ -140,6 +140,12 @@ static_assert(
     "Failed testing variables_tag_with_tags_list");
 /// [variables_tag_with_tags_list]
 
+static_assert(
```
A test with 2 tags would be a good addition (though unlikely to be different, 1 is a fair special case) :)

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -397,8 +397,27 @@
  * \details Actions and parallel components may require an elliptic system to
```
In the commit message `allows elliptic code` -> `allows the elliptic code`

nilsdeppe reviewed (commented):
A couple minor things, I'm happy for you to squash :)

wthrowe commented on src/Evolution/Actions/ComputeTimeDerivative.hpp:
```diff
@@ -0,0 +1,81 @@
+/*!
+ * \ingroup ActionsGroup
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Compute the bulk contribution to the time derivative of the variables
```
Reword to not assume DG.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyFluxes.hpp:
```diff
@@ -59,27 +74,23 @@ struct ApplyBoundaryFluxesGlobalTimeStepping {
     constexpr size_t volume_dim = system::volume_dim;
     using variables_tag = typename system::variables_tag;
-    using dt_variables_tag = db::add_tag_prefix<Tags::dt, variables_tag>;
+    using dt_variables_tag =
```
Do you want to rename this?  (Up to you.)

nilsvu replied:
I think it's a fine name for this local variable. Didn't want to make it too abstract.

wthrowe commented on tests/Unit/Evolution/Actions/Test_ComputeTimeDerivative.cpp:
```diff
@@ -21,12 +21,17 @@
 };
 
-struct ComputeDuDt {
+struct ComputeVolumeDuDt {
```
Why this change?

nilsvu replied:
left-over from the name changes. reverted

nilsvu commented:
@wthrowe @nilsdeppe @kidder squashed and rebased

nilsdeppe reviewed (approved):
Restarted GCC debug builds
